### PR TITLE
Flame graph percentage display

### PR DESCRIPTION
**Pull request by iamstuartwilson**
## Overview

 - Adds a dropdown inside the info box which displays stack percentages (top and overall)
 - Centres text inside info box
 - Stacks text inside info box on smaller screens with ability to scroll to see all text
 - Adds titles to each text element inside info box in case of massive path or function string

## Before
https://upload.clinicjs.org/public/c748ff5b367a00825f2f8157d85d70c54402cf628b02faa5ca9437a92e63bcb3/49653.clinic-flame.html#selectedNode=207&zoomedNode=&exclude=8000&merged=true

## After
https://upload.clinicjs.org/public/eb6802c7c59ef7005c46ae87e42d1d46fc7f2693bbeb099a8396ea945755c0e4/49653.clinic-flame.html#selectedNode=207&zoomedNode=&exclude=8000&merged=true

### CODE REVIEW

**Comment by mcollina** [comment]
Good work!

**Comment by AlanSl** [comment]
Nice. Reviewing the output - general approach looks good, one immediate issue, the sample output is running into errors on V8 Runtime frames for some reason. UI shows 0% and interacting with them causes console errors:

image

**Comment by AlanSl** [comment]
A couple of small tweaks with the new info box styling (there may be some more):

1. Can we make sure there's a gap between function name, file name etc when they collide, for example this could be misread as the .. being part of the function name:

image

https://upload.clinicjs.org/public/08af474a186a7706657249393ff883d48bbf7bc7eb24db3b2634344ce3eb8525/2010.clinic-flame.html#selectedNode=1234&zoomedNode=&exclude=83c0&merged=true

2. On medium-narrow screens with scrollbars visible, it can become more scrollbar than content when something is long, like this below. Maybe turn off white-space: nowrap below the breakpoint that stacks the text, or similar.

image

3. If the text extends beyond 3 lines, the top lines extend beyond the top of the scrollable area and become unreachable (I told you this one was tricky! 🙃 Feel free to move it to a separate PR so that tweaking and reviewing it doesn't snarl up the %s feature)

image

https://upload.clinicjs.org/public/7659886ad47de313f68383dcfd350142e6405d5e44367c24fa33a6c68fb1c639/41533.clinic-flame.html#selectedNode=116&zoomedNode=&exclude=8000&merged=true

Regarding the content of the box, I'd like to tweak it to add some expandable explanatory text and "X of Y samples" but it's probably easiest to do that as another PR on this one.

**Comment by iamstuartwilson** [comment]
@AlanSl I've nuked the changes to the info box text layout and retained the dropdown. Updated here: https://upload.clinicjs.org/public/f19a12ec5b3830a452daa3fe38d7ff5b90505805625cfb214b8d23523d531d41/49653.clinic-flame.html#selectedNode=64&zoomedNode=&exclude=83c0&merged=true

**Comment by AlanSl** [comment]
Can you at some point post those changes in a separate PR? They looked promising, just needing some tricky fixes and whack-a-mole. 

Looks like you've fixed those errors 

One little glitch immediately jumping out in that demo: no gap between the file path and the line:column numbers. I was going to say before, can we keep the longer/double space there so that the file path is easy to skim (easy to land your eye on the last part i.e. the file name itself)

image

**Comment by iamstuartwilson** [comment]
@AlanSl goos post. To save the server to much bloat here's a visual diff between master and this branch with the latest glitch fix:

info-updates

**Comment by AlanSl** [comment]
Nice, (and 👍 for the gif!) but can we keep the single spacing _within_ `line:X column:Y`, so at a glance it's two easy to distinguish units not three (a filename, where it's easy to skim to the end, and line-and-column details)

**Comment by iamstuartwilson** [comment]
@AlanSl updated. Sorry, human error 👨‍💻

**Comment by iamstuartwilson** [comment]
@AlanSl done and thanks! A separate, holistic UI PR would be great to review consistency 👍

**Comment by cuginoAle** [comment]
Looking great!
Should the dropdown auto-collapse when clicking anywhere else on the page? (same behavior of Options drop down)

File: visualizer/data-tree.js [hunk 1]
```diff
@@ -30,6 +30,9 @@ class DataTree {
     this.maxRootAboveMean = 0
     this.maxRootBelowMean = 0
 
+    // Set by countTotalFrames
```
Comment (cuginoAle) [inline visualizer/data-tree.js]: Hmm... I can't see where `countTotalFrames` is defined

**Comment by cuginoAle** [comment]
I left just a couple of minor comments.
Other than that it LGTM!

File: visualizer/data-tree.js (hunk 1 above)
Comment (iamstuartwilson) [reply visualizer/data-tree.js]: Good spot, a hangover from a different approach 👍

**Comment by iamstuartwilson** [comment]
@cuginoAle I've addressed those two points. Thanks a lot for taking the time to review!

**Comment by iamstuartwilson** [comment]
@AlanSl @cuginoAle would be good to double check this works still with the UI updates and whether it might conflict with other percentage work.

I've rebased and uploaded a profile here: https://preprod.clinicjs.org/public/ee6adfc257d89e99723d625ca321d3f04f112156b86b712f2aa298acee2048f0/5338.clinic-flame.html#selectedNode=946&zoomedNode=&exclude=83c0&merged=true

If all looks good to you still, might be nice to get this out since it's been hanging around a while :)

**Comment by cuginoAle** [comment]
I noticed the following z-index issue:
screenshot 2019-02-22 at 09 41 07

I believe the bug is not related to this PR though, it might have been there since day 1 and we didn't notice it before ;)

**Comment by AlanSl** [comment]
We're creating a `dev` branch for https://github.com/nearform/node-clinic-flame/pull/85 and related new features - when that PR is merged into `dev` can you make this point against `dev` instead of `master` and update it, moving the %s box to the new space on the right-hand side, then we can (finally!) merge it

**Comment by iamstuartwilson** [comment]
> We're creating a `dev` branch for #85 and related new features - when that PR is merged into `dev` can you make this point against `dev` instead of `master` and update it, moving the %s box to the new space on the right-hand side, then we can (finally!) merge it.
> 
> And then the same with #75

Awesome! Will do @AlanSl 👍

**Comment by iamstuartwilson** [comment]
@AlanSl @cuginoAle I've moved the dropdown into the space freed up by the footer work. Although it's not longer part of the info-box visually, the info-box hooks into a lot of functionality which made it more suitable to living within that element, even though stylistically it sits inline with the selection controls.

Updated here: https://preprod.clinicjs.org/public/3be33843bfda13bd4083c9cd723a89862618c11a0d2d42594e1773497273b153/5338.clinic-flame.html#selectedNode=938&zoomedNode=&exclude=83c0&merged=true

**Comment by AlanSl** [comment]
Tiny tweak: can you make the height and vertical alignment of the text match the buttons on the left? Then good to go

**Comment by iamstuartwilson** [comment]
My mistake, was sure I'd fixed that 🤔

Updated: https://preprod.clinicjs.org/public/f29ad0a0b7cd0028451afe52536e6a725b5da36a87fcbd1c94225bbcef5cff81/5338.clinic-flame.html#selectedNode=938&zoomedNode=&exclude=83c0&merged=true

**Pull request merged**